### PR TITLE
SquashFS use zstd as default

### LINKED ISSUE #176: SquashFS picking a compression algorithm
When I implemented SquashFS-support for conda-pack I only did some rudimentary tests regarding the runtime of packing & unpacking with the different compression options of `mksquashfs`.

I finally got around to running slightly more detailed performance tests. 
I ran the perftests on a project that I had lying around with a conda env of size ~1GB:
1. `conda-pack` the environment to SquashFS
2. Mount the packed environment in writeable mode using squashfs + overlayfs.
3. `source mountpoint/activate && conda-unpack && pip install -e . && pytest && source mountpoint/deactivate`.

| Compression | Archive size in MB | runtime `conda-pack` | Runtime activate & install & pytest |
| --- | --- | --- | --- | 
| no compression | 890 | 0:05 | 0:57 |
| gzip | 263 | 0:24 | 0:59 |
| zstd -Xcompression-level 1 | 286 | 0:05 | 0:58 |
| zstd -Xcompression-level 1 -b 262144 -noI -noX| 279 | 0:05 | 0:58 |
| zstd -Xcompression-level 8 | 251 | 0:10 | 0:57 |
| zstd -Xcompression-level 8 -b 262144 -noI -noX| 244 | 0:12 | 0:57 |
| zstd -Xcompression-level 16 | 238 | 0:45 | 0:57 |
| zstd -Xcompression-level 16 -b 262144 -noI -noX | 230 | 0:49 | 0:57 |
| zstd -Xcompression-level 16 -b 262144 | 228 | 0:50 | 0:57 |
| zstd -Xcompression-level 16 -noI -noX | 239 | 0:45 | 0:57 |
| zstd -Xcompression-level 16 -noI | 239 | 0:43 | 0:56 |
| zstd -Xcompression-level 16 -noX | 238 | 0:44 | 0:57 |
| zstd -Xcompression-level 22 | 236 | 1:54 | 0:57 |
| zstd -Xcompression-level 22 -b 262144 -noI -noX | 229 | 1:35 | 0:57 |
| lz4 | 387 | 0:05 | 0:56 |
| lz4 -b 262144 -noI -noX | 385 | 0:05 | 0:56 |
| xz | 219 | 0:54 | 1:07 |
| no squashFS (tar.gz) | 280 | 0:11 | 1:01 |
| no conda-pack | 980 | 0:00 | 0:54 |

- `noI`: don't compress inode table
- `noX`: Don't compress extended attrs

For reference, [Xar](https://github.com/facebookincubator/xar/blob/a6f15d1664cb1171d5b8e7e05a20dbd8b2ceb652/xar/xar_util.py#L87) uses `zstd -Xcompression-level 16 -b 262144 -noI -noX`. `libzstd` comes pre-installed on Ubuntu and Fedora, it's on brew and conda-forge, too.

Unless the maintainers think otherwise I'll open a PR to use `zstd` with varying levels of `-Xcompression-level` based on `conda-pack --compress-level`, only falling back to `xz` at the highest level. `zstd -Xcompression-level 8` seems like a sane default.

### PR DESCRIPTION
**Pull request by siboehm**
Closes #176

### CODE REVIEW

**Comment by dbast** [comment]
@siboehm This looks really good. Thanks for investigating and polishing the compression settings.

Out of scope of this PR, but still in the context of polishing the SquashFS feature: Would it make sense to also add a check to the constructor of SquashFSArchive to check that the mksquashfs command is available with a message ("Please make sure `mksquashfs` is available, e.g. install it via `conda install squashfs-tools`) and exit ... or maybe instead handle the check_call call FileNotFoundError exception? That would be great from user perspective ... And that is better than adding squashfs-tools as run dependencies, as those are not available on pypi and on conda-forge not for all platforms.

**Comment by dbast** [comment]
This PR has one drawback: While previously one could select between gzip, lzo and xz as squashfs compression algorithm by choosing different levels, it is then only zstd and xz. During creation that is maybe not a problem as e.g. the conda-forge::squashfs-tools are compiled with gzip, lzo, lz4, zstd and xz activated. But when trying to mount the conda-packed environment on another machine the existing kernel (afaik kernel >=4.14) or squashfuse have to be recent enough.

**Comment by siboehm** [comment]
That's true. We need `squashfuse >=1.101` (released 12.2017) or kernel `>=4.14` (released 11.2017) for `zstd`. However with `xz` we still have a compression option that works as far back as Kernel 2.6 (03.2011). 

I don't want to change the API and I want it to be easy for users to get good performance without having to tweak anything. We could introduce a `gzip` option somewhere in the `compression_level` hierarchy, however I'd like to keep the default as `zstd`. Personally I think xz and squashfuse are enough options for working around an old kernel.

**Comment by dbast** [comment]
Looking at current/LTS versions of still supported Linux distributions (Debian 9/10, CentOS 7/8, Ubuntu 18.04/20.04), shows that not all have squashfuse version with zstd support, e.g Debian 10 has version 0.1.103 but compiled without zstd support, Ubuntu 18.04 has 0.1.100 without zstd. Also the Linux kernels of Debian 9 and CentOS 7 are potentially (if nothing got backported) too old.

I understand that having zstd as default would be technically super cool, while there are still tons of conservative managed machines running somewhere with older versions.

Long story short prevent any surprises, why not present the full information to the user by adding e.g. a print statement in case of zstd (the default) with "Using zstd level {} for best compression / speed ratio, which requires kernel >=4.14 or squashfuse >=0.1.101 compiled with zstd support for mounting (Select `--compress-level 9`, which uses xz, for compatibility with older versions.)"?

**Comment by siboehm** [comment]
Ok I tried to be explicit about the downside of zstd in the Documentation, plus I added a log message.

File: conda_pack/formats.py [hunk 1]
```diff
@@ -363,6 +363,10 @@ def zipinfo_from_file(filename, arcname=None):
 class SquashFSArchive(ArchiveBase):
     def __init__(self, fileobj, target_path, arcroot, n_threads, verbose=False,
                  compress_level=4):
+        if shutil.which("mksquashfs") is None:
+            print("Command 'mksquashfs' not found. Please install it, "
+                  "e.g. 'conda install squashfs-tools'.")
+            sys.exit(127)
```
Comment (siboehm) [inline conda_pack/formats.py:369]: This works, but doesn't clean up the `temp_path` created here: https://github.com/conda/conda-pack/blob/4ad334d4338c12bb81b108afbf1116a535e1478b/conda_pack/core.py#L367

Instead raise an Exception, but which a helpful error msg, which will cleanup the path automatically? 

I guess it's pretty rare, so not super important.

Comment (dbast) [reply conda_pack/formats.py]: Thanks for catching. Fixed and tested that tempfiles are really cleaned.

**Review by dbast — APPROVED** [review]

**Comment by dbast** [comment]
@siboehm Thanks!
